emma-language/src/main/scala/org/emmalanguage/compiler/lang/FunSupport.scala

```diff
@@ -0,0 +1,478 @@
+
+        def assign(u: CG.Vertex, r: CG.Vertex): Unit =
+          if (!component.contains(u)) {
+            component(u) = component.getOrElse(r, r)
+            in(u).foreach(v => assign(v, r))
+          }
+
+        // the actual algorithm
+        for {
+          u <- postorder()
+        } assign(u, u)
+
+        component.toMap
+      }
+
+      val assignment = assign()
+
+      // transform the component map into a nested set of strongly connected components
+      assignment.toSet[(CG.Vertex, CG.Vertex)]
+        .groupBy { case (v, c) => c }
+        .values.toSet[Set[(CG.Vertex, CG.Vertex)]]
+        .map(_.map { case (v, c) => v })
+    }
+
+    // -------------------------------------------------------------------------
+    // Helper methods & objects
+    // -------------------------------------------------------------------------
+
+    /** Build a map of all lambdas defined in the given tree. */
+    private lazy val lambdaDefsIn: u.Tree => Map[u.TermSymbol, CG.Lambda] = _.collect({
+      case ValDef(sym, lambda@Lambda(_, _, _), _) => sym -> CG.Lambda(sym, lambda)
+    }).toMap
+
+    /** Build a map of all library functions called in the given tree. */
+    private lazy val libdefRefsIn: u.Tree => Map[u.TermSymbol, CG.LibDef] = _.collect({
+      case DefCall(Some(TermRef(tgt)), sym withAST ast, targs, argss@_*)
+        if get.ann[emma.src](sym).isDefined => sym -> CG.LibDef(sym, ast)
+    }).toMap
+
+    /* Lirary function registry. */
+    object LibDefRegistry {
+
+      /** If the given MethodSymbol is a library function, loads it into the registry and retuns its AST. */
+      def apply(sym: u.MethodSymbol): Option[u.DefDef] = getOrload(sym)
+
+      /** Extractor for the AST associated with a library function symbol. */
+      object withAST {
+        def unapply(sym: u.MethodSymbol): Option[(u.MethodSymbol, u.DefDef)] =
+          apply(sym).map(ast => (sym, ast))
+      }
+
+      /** Extractor for the `emma.src` annotation associated with a library function symbol. */
+      object withAnn {
+        def unapply(sym: u.MethodSymbol): Option[(u.MethodSymbol, u.Annotation)] =
+          get.ann[emma.src](sym) match {
+            case Some(ann) => Some(sym, ann)
+            case None => None
+          }
+      }
+
+      /** Loads and memoizes an AST for a library function from a `sym` with an `emma.src` annotation. */
+      private lazy val getOrload = Memo[u.MethodSymbol, Option[u.DefDef]]({
+        case sym withAnn ann =>
+          ann.tree.collect {
+            case Lit(fldName: String) =>
+              // build an `objSym.fldSym` selection expression
+              val cls = sym.owner.asClass
+              val fld = get.tpe(cls).member(TermName(fldName)).asTerm
+              val sel = qualifyStatics(DefCall(Some(Ref(cls.module)))(fld)())
+              // evaluate `objSym.fldSym` and grab the DefDef source
+              val src = eval[String](unTypeCheck(sel))
```

Comment:
`unTypeCheck` is not needed here, as we also do it inside `eval`. (see the note in the scaladoc of `eval`)